weather_dl/download_pipeline/manifest.py

```diff
@@ -242,22 +395,141 @@ def _update(self, download_status: DownloadStatus) -> None:
+                                 "if a point lies in the given area or not."),
+            bigquery.SchemaField('stage', 'STRING', mode='NULLABLE',
+                                 description="Current stage of request : 'fetch', 'download', 'retrieve', 'upload' "
+                                 "or None."),
+            bigquery.SchemaField('status', 'STRING', mode='REQUIRED',
+                                 description="Download status: 'scheduled', 'in-progress', 'success', or 'failure'."),
+            bigquery.SchemaField('error', 'STRING', mode='NULLABLE',
+                                 description="Cause of error, if any."),
+            bigquery.SchemaField('user', 'STRING', mode='REQUIRED',
+                                 description="Identifier for the user running the download."),
+            bigquery.SchemaField('size', 'FLOAT', mode='NULLABLE',
+                                 description="Shard size in GB."),
+            bigquery.SchemaField('scheduled_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when download was scheduled."),
+            bigquery.SchemaField('retrieve_start_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the retrieve stage starts."),
+            bigquery.SchemaField('retrieve_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the retrieve state ends."),
+            bigquery.SchemaField('fetch_start_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the fetch state starts."),
+            bigquery.SchemaField('fetch_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the fetch state ends."),
+            bigquery.SchemaField('download_start_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the download state starts."),
+            bigquery.SchemaField('download_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the download state ends."),
+            bigquery.SchemaField('upload_start_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the upload state starts."),
+            bigquery.SchemaField('upload_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the upload state ends."),
+        ]
+        table = bigquery.Table(self.location, schema=TABLE_SCHEMA)
+        with bigquery.Client() as client:
+            client.create_table(table, exists_ok=True)
+
+    def _read(self, location: str) -> DownloadStatus:
+        """Reads the JSON data from a manifest."""
+        with bigquery.Client() as client:
+            query = f"SELECT * FROM {self.location} WHERE location = {location!r}"
```

Comment:
Can you sanitize these inputs too?

Reply:
I think this will only be necessary if the `query()` command can write or delete data in the DB. 

Reading the docs, it looks like this is the case.

https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.client.Client#google_cloud_bigquery_client_Client_query

Reply:
Done. Thanks for pointing this out.